operator/pkg/inneroperator/values/translation/translation.go

```diff
@@ -0,0 +1,442 @@
+package translation
+
+import (
+	"context"
+	"crypto/sha256"
+	"crypto/tls"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"strconv"
+
+	// Required for the usage of go:embed below.
+	_ "embed"
+
+	"github.com/hashicorp/go-multierror"
+	"github.com/pkg/errors"
+	"github.com/stackrox/rox/generated/storage"
+	platform "github.com/stackrox/rox/operator/apis/platform/v1alpha1"
+	"github.com/stackrox/rox/operator/pkg/securedcluster/scanner"
+	"github.com/stackrox/rox/operator/pkg/values/translation"
+	helmUtil "github.com/stackrox/rox/pkg/helm/util"
+	"github.com/stackrox/rox/pkg/pointers"
+	"github.com/stackrox/rox/pkg/utils"
+	"go.uber.org/zap"
+	"helm.sh/helm/v3/pkg/chartutil"
+	corev1 "k8s.io/api/core/v1"
+	k8sErrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrlClient "sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/yaml"
+)
+
+const (
+	sensorTLSSecretName           = "sensor-tls"
+	admissionControlTLSSecretName = "admission-control-tls"
+	collectorTLSSecretName        = "collector-tls"
+
+	legacyCollectionKernelModule = "KernelModule"
+)
+
+var (
+	//go:embed base-values.yaml
+	baseValuesYAML []byte
+)
+
+// New creates a new Translator
+func New(client ctrlClient.Client) Translator {
+	return Translator{client: client}
+}
+
+// Translator translates and enriches helm values
+type Translator struct {
+	client ctrlClient.Client
+}
+
+// Translate translates and enriches helm values
+func (t Translator) Translate(ctx context.Context, u *unstructured.Unstructured) (chartutil.Values, error) {
+	baseValues, err := chartutil.ReadValues(baseValuesYAML)
+	utils.CrashOnError(err) // ensured through unit test that this doesn't happen.
+
+	sc := platform.SecuredCluster{}
+	err = runtime.DefaultUnstructuredConverter.FromUnstructured(u.Object, &sc)
+	if err != nil {
+		return nil, err
+	}
+
+	valsFromCR, err := t.translate(ctx, sc)
+	if err != nil {
+		return nil, err
+	}
+
+	logger, _ := zap.NewProduction()
+	sugar := logger.Sugar()
+	targetVersion, err := t.queryCentralTargetVersion(valsFromCR)
+	if err != nil {
+		return nil, err
+	}
+
+	tversion, err := helmUtil.ValuesForKVPair("operator.version", targetVersion)
+	if err != nil {
+		return nil, err
+	}
+	sugar.Infof("tversion: %v", tversion)
+	// Overwrite target image version of Operator with result from central
+	// CHECKME: As this is preferred left-to-right, is this a bug in the other implementations?
+	//			Technically, we would need to apply CRs first, then imageOverrides, then baseValues, no?
+	values := helmUtil.CoalesceTables(tversion, baseValues)
```

Comment:
Is it intended that the `valsFromCR` are not returned by the `Translate` function?

Reply:
Yes, that is correct. At this point in time, we are only interested in the `centralEndpoint` of the Secured Cluster CR.
As we are in Outer Operator Mode, we mainly process Central CRs fully.
We listen passively to the creation of Secured Cluster CRs, extract only the central endpoint from it and query that endpoint for a target version of the inner Operator to deploy.